circuits/cpp/src/aztec3/circuits/hash.hpp

```diff
@@ -269,4 +271,33 @@ template <typename NCT> typename NCT::fr compute_public_data_tree_index(typename
     return NCT::compress({ contract_address, storage_slot }, GeneratorIndex::PUBLIC_LEAF_INDEX);
 }
 
+template <typename NCT> typename NCT::fr compute_l2_to_l1_hash(typename NCT::address contract_address,
+                                                               typename NCT::fr rollup_version_id,
+                                                               typename NCT::fr portal_contract_address,
+                                                               typename NCT::fr chain_id,
+                                                               typename NCT::fr content)
+{
+    using fr = typename NCT::fr;
+
+    std::vector<fr> const inputs = {
+        contract_address.to_field(), rollup_version_id, portal_contract_address, chain_id, content,
+    };
+
+    constexpr auto const num_bytes = 5 * 32;
+    std::array<uint8_t, num_bytes> calldata_hash_inputs_bytes;
+    // Convert all into a buffer, then copy into the array, then hash
+    for (size_t i = 0; i < inputs.size(); i++) {
+        auto as_bytes = inputs[i].to_buffer();
+
+        auto offset = i * 32;
+        std::copy(as_bytes.begin(), as_bytes.end(), calldata_hash_inputs_bytes.begin() + offset);
+    }
```

Comment:
I don't recall why it was done like this, think I just took what we had in the components 😅, so both might need an update.

Reply:
Not sure if this is relevant to the topic raised Sean, but we would still need to still pad them to 32 bytes each. In solidity, before hashing, both L1Actor abd L2Actor is 64 bytes each (since they have 2 fields each) after doing `abi.encode`

Reply:
Ok, I've made a draft issue in a3 faries to refactor this